core/dmux.go

```diff
@@ -311,10 +358,204 @@ func simpleSetup(size, qsize int, sink Sink) ([]chan interface{}, *sync.WaitGrou
+	sidelineChannel []chan ChannelObject, sinkChannel []chan ChannelObject, wg *sync.WaitGroup) {
+	for msg := range ch[index] {
+		key := source.GetKey(msg)
+		partition := source.GetPartition(msg)
+		value := source.GetValue(msg)
+		offset := source.GetOffset(msg)
+		var check sideline_models.CheckMessageSidelineResponse
+		expBackOff := backoff.NewExponentialBackOff()
+		//expBackOff.MaxElapsedTime = math.MaxInt32 * time.Minute
+		retryError := backoff.Retry(func() error {
+			log.Printf("Checking if the message is already sidelined %d, %d from channel", partition, offset)
+			checkSidelineMessage := sideline_models.SidelineMessage{
+				GroupId:           string(key),
+				Partition:         partition,
+				EntityId:          string(key) + sideline.ConsumerGroupName + sideline.ClusterName,
+				Offset:            offset,
+				ConsumerGroupName: sideline.ConsumerGroupName,
+				ClusterName:       sideline.ClusterName,
+				Message:           value,
+				ConnectionType:    sideline.ConnectionType,
+			}
+			checkSidelineMessageBytes, checkSidelineMessageErr := json.Marshal(checkSidelineMessage)
+			if checkSidelineMessageErr != nil {
+				log.Printf("error in serde of checkSidelineMessage " + checkSidelineMessageErr.Error())
+				return errors.New("error in serde of checkSidelineMessage " + checkSidelineMessageErr.Error())
+			}
+			checkBytes, checkErr := sidelineImpl.CheckMessageSideline(checkSidelineMessageBytes)
+			err := json.Unmarshal(checkBytes, &check)
+			if err != nil {
+				log.Printf("error in serde of CheckMessageSidelineResponse " + err.Error())
+				return errors.New("error in serde of CheckMessageSidelineResponse " + err.Error())
+			}
+			if checkErr != nil {
+				log.Printf("Error in checking if message is sidelined " + checkErr.Error())
+				return errors.New("Error in checking if message is sidelined " + checkErr.Error())
+			}
+			log.Printf("Message if already sidelined %t %d %d", check.SidelineMessage, partition, offset)
+			if check.SidelineMessage {
+				sendToSidelineChannel := ChannelObject{
+					Msg:      msg,
+					Sideline: sideline,
+					Version:  check.Version,
+				}
+				sidelineChannel[index] <- sendToSidelineChannel
+				return nil
+			} else {
+				sendToSinkChannel := ChannelObject{
+					Msg:      msg,
+					Sideline: sideline,
+				}
+				sinkChannel[index] <- sendToSinkChannel
+				return nil
+			}
+		}, expBackOff)
+		if retryError != nil {
+			log.Fatal("Ideally this should not happen in mainChannelConsumption")
+		}
+	}
+	wg.Done()
+}
+
+func pushToSideline(sidelineChannel []chan ChannelObject, index int, source Source, sideline Sideline, sidelineImpl sideline_models.CheckMessageSideline) {
+	for channelObject := range sidelineChannel[index] {
+		expBackOff := backoff.NewExponentialBackOff()
+		//expBackOff.MaxElapsedTime = math.MaxInt32 * time.Minute
+		retryError := backoff.Retry(
+			func() error {
+				sidelineMetaByteArray, sidelineMetaByteArrayErr := json.Marshal(sideline.SidelineMeta)
```

Comment:
why is this failure so late in the funnel? and why do we need to deserialize every time we proess message in channel?

Reply:
changed this